File: user-web/src/api/algorithmReset.ts
import { http } from "./http";

export type AlgorithmState = {
  mode: "origin" | string;
  personalized: boolean;
  resetAt: string | null;
  explorationSeed: string | null;
  coldStartStrategy: "random_balanced" | string;
  parameters: AlgorithmParameters;
  attentionIndex: number;
};

export type AlgorithmParameters = {
  randomness: number;
  diversity: number;
  depth: number;
  entertainment: number;
  challenge: number;
  novelty: number;
};

export type AlgorithmResetResponse = {
  success: boolean;
  message: string;
  data: AlgorithmState;
};

export async function algorithmReset(): Promise<AlgorithmResetResponse> {
  const { data } = await http.post<AlgorithmResetResponse>("/users/me/algorithm-reset");
  return data;
}

export async function getAlgorithmState(): Promise<AlgorithmResetResponse> {
  const { data } = await http.get<AlgorithmResetResponse>("/users/me/algorithm-state");
  return data;
}

export async function updateAlgorithmState(body: {
  mode: "origin" | "efficiency" | "growth" | "emotion" | "custom";
  parameters?: AlgorithmParameters;
}): Promise<AlgorithmResetResponse> {
  const { data } = await http.patch<AlgorithmResetResponse>("/users/me/algorithm-state", body);
  return data;
}

export type AlgorithmPreset = {
  id: string;
  name: string;
  description: string | null;
  parameters: AlgorithmParameters;
  createdAt: string;
  updatedAt: string;
};

export async function listAlgorithmPresets(): Promise<{ success: boolean; items: AlgorithmPreset[] }> {
  const { data } = await http.get("/users/me/algorithm-presets");
  return data;
}

export async function saveAlgorithmPreset(body: {
  name: string;
  description?: string | null;
  parameters?: AlgorithmParameters;
}): Promise<{ success: boolean; message: string; data: AlgorithmPreset }> {
  const { data } = await http.post("/users/me/algorithm-presets", body);
  return data;
}

export async function applyAlgorithmPreset(id: string): Promise<AlgorithmResetResponse> {
  const { data } = await http.post<AlgorithmResetResponse>(`/users/me/algorithm-presets/${id}/apply`);
  return data;
}

export async function resetCustomDefaults(): Promise<AlgorithmResetResponse> {
  const { data } = await http.post<AlgorithmResetResponse>("/users/me/algorithm-state/reset-custom-defaults");
  return data;
}

export async function getAttentionIndex(): Promise<{
  success: boolean;
  data: {
    attentionIndex: number;
    inputQuality: number;
    outputValue: number;
    cognitiveGrowth: number;
    timeQuality: number;
    informationValue: number;
    propagationImpact: number;
    deepEngagement: number;
    formula: string;
    detail: Record<string, unknown> | null;
    latestMode: string | null;
    explanation: string;
  };
}> {
  const { data } = await http.get("/users/me/attention-index");
  return data;
}

export type AiAgentPanel = {
  mode: string;
  modelVersion: string;
  summary: string;
  contentSummary: string;
  attentionOptimization: string;
  learningPathSuggestion: string;
  alerts: string[];
  attentionFactors: Record<string, number>;
  suggestions: string[];
  nextActions: string[];
};

export async function getAiAgentPanel(): Promise<AiAgentPanel> {
  const { data } = await http.get("/users/me/ai-agent-panel");
  return data;
}

export async function getGovernancePower(): Promise<{ votingPower: number; formula: string; canVote: boolean }> {
  const { data } = await http.get("/users/me/governance-power");
  return data;
}

File: backend/app/schemas/algorithm_reset.py
from __future__ import annotations

from datetime import datetime
import uuid

from pydantic import BaseModel, Field

AlgorithmMode = str
ATTENTION_VALUE_FORMULA = "Attention Value = 时间质量 × 信息价值 × 传播影响 × 深度参与"


class AlgorithmParameters(BaseModel):
    randomness: int = Field(default=50, ge=0, le=100)
    diversity: int = Field(default=70, ge=0, le=100)
    depth: int = Field(default=50, ge=0, le=100)
    entertainment: int = Field(default=50, ge=0, le=100)
    challenge: int = Field(default=50, ge=0, le=100)
    novelty: int = Field(default=60, ge=0, le=100)


class AlgorithmStateOut(BaseModel):
    mode: str
    personalized: bool
    resetAt: datetime | None = None
    explorationSeed: str | None = None
    coldStartStrategy: str
    parameters: AlgorithmParameters = Field(default_factory=AlgorithmParameters)
    attentionIndex: int = 0


class AlgorithmResetResponse(BaseModel):
    success: bool
    message: str
    data: AlgorithmStateOut


class AlgorithmStateUpdate(BaseModel):
    mode: AlgorithmMode = Field(pattern="^(origin|efficiency|growth|emotion|custom)$")
    parameters: AlgorithmParameters | None = None


class AlgorithmModeOption(BaseModel):
    mode: str
    title: str
    subtitle: str
    weights: AlgorithmParameters


class AlgorithmStateResponse(BaseModel):
    success: bool
    message: str
    data: AlgorithmStateOut


class AlgorithmPresetCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    description: str | None = Field(default=None, max_length=240)
    parameters: AlgorithmParameters | None = None


class AlgorithmPresetOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None = None
    parameters: AlgorithmParameters
    createdAt: datetime
    updatedAt: datetime


class AlgorithmPresetListResponse(BaseModel):
    success: bool
    items: list[AlgorithmPresetOut]


class AlgorithmPresetResponse(BaseModel):
    success: bool
    message: str
    data: AlgorithmPresetOut


class AttentionIndexOut(BaseModel):
    attentionIndex: int
    inputQuality: int
    outputValue: int
    cognitiveGrowth: int
    timeQuality: int = 0
    informationValue: int = 0
    propagationImpact: int = 0
    deepEngagement: int = 0
    formula: str = ATTENTION_VALUE_FORMULA
    detail: dict | None = None
    latestMode: str | None = None
    explanation: str


class AttentionIndexResponse(BaseModel):
    success: bool
    data: AttentionIndexOut


class AiAgentPanelOut(BaseModel):
    mode: str
    modelVersion: str = "agent-v1-rule-based"
    summary: str
    contentSummary: str
    attentionOptimization: str
    learningPathSuggestion: str
    alerts: list[str] = Field(default_factory=list)
    attentionFactors: dict[str, int] = Field(default_factory=dict)
    suggestions: list[str]
    nextActions: list[str]


class GovernancePowerOut(BaseModel):
    votingPower: int
    formula: str
    canVote: bool
